worker/lease/store.go

```diff
@@ -0,0 +1,422 @@
+	}
+}
+
+// Leases (lease.Store) returns all leases in the database,
+// optionally filtering using the input keys.
+func (s *Store) Leases(keys ...lease.Key) (map[lease.Key]lease.Info, error) {
+	// TODO (manadart 2022-11-30): We expect the variadic `keys` argument to be
+	// length 0 or 1. It was a work-around for design constraints at the time.
+	// Either filter the result here for len(keys) > 1, or fix the design.
+	// As it is, there are no upstream usages for more than one key,
+	// so we just lock in that behaviour.
+	if len(keys) > 1 {
+		return nil, errors.NotSupportedf("filtering with more than one lease key")
+	}
+
+	name := "Leases"
+	q := `
+SELECT t.type, l.model_uuid, l.name, l.holder, l.expiry
+FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id`[1:]
+
+	var args []any
+
+	if len(keys) == 1 {
+		q += `
+WHERE  t.type = ?
+AND    l.model_uuid = ?
+AND    l.name = ?`
+
+		name = "LeasesForKey"
+		key := keys[0]
+		args = []any{key.Namespace, key.ModelUUID, key.Lease}
+	}
+
+	stmt, err := s.getPrepared(context.Background(), name, q)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result, err := leasesFromRows(rows)
+	return result, errors.Trace(rows.Err())
+}
+
+// ClaimLease (lease.Store) claims the lease indicated by the input key,
+// for the holder and duration indicated by the input request.
+// The lease must not already be held, otherwise an error is returned.
+func (s *Store) ClaimLease(key lease.Key, req lease.Request, stop <-chan struct{}) error {
+	if err := req.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+INSERT INTO lease (uuid, lease_type_id, model_uuid, name, holder, start, expiry)
+SELECT ?, id, ?, ?, ?, datetime('now'), datetime('now', ?) 
+FROM   lease_type
+WHERE  type = ?`[1:]
+
+		stmt, err := s.getPrepared(ctx, "ClaimLease", q)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		d := fmt.Sprintf("+%d seconds", int64(math.Ceil(req.Duration.Seconds())))
+
+		_, err = stmt.ExecContext(
+			ctx, utils.MustNewUUID().String(), key.ModelUUID, key.Lease, req.Holder, d, key.Namespace)
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.Trace(<-errCh)
+	case err := <-errCh:
+		cancel()
+		if database.IsErrConstraintUnique(err) {
+			return lease.ErrHeld
+		}
+		return errors.Trace(err)
+	}
+}
+
+// ExtendLease (lease.Store) ensures the input lease will be held for at least
+// the requested duration starting from now.
+// If the input holder does not currently hold the lease, an error is returned.
+func (s *Store) ExtendLease(key lease.Key, req lease.Request, stop <-chan struct{}) error {
+	if err := req.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This should be obsolete when we start passing context.Context through the API server stack.